File: py-radiate/radiate/limit.py
from radiate.radiate import PyLimit


class LimitBase:
    def __init__(self, limit: PyLimit):
        """
        Initialize the limit base with a Limit instance.
        :param limit: An instance of Limit.
        """
        self.limit = limit

    def __str__(self):
        """
        Return a string representation of the limit.
        :return: String representation of the limit.
        """
        return f"Limit(name={self.limit.name}, args={self.limit.args})"

    def __repr__(self):
        """
        Return a detailed string representation of the limit.
        :return: Detailed string representation of the limit.
        """
        return f"LimitBase(limit={self.limit})"


class SecondsLimit(LimitBase):
    """
    Limit the execution time of the engine.
    """

    def __init__(self, seconds: int):
        """
        Initialize the seconds limit.
        :param seconds: Number of seconds to limit the execution time.
        """
        super().__init__(limit=PyLimit.Seconds(seconds))


class ScoreLimit(LimitBase):
    """
    Limit the score of the engine.
    """

    def __init__(self, score: float):
        """
        Initialize the score limit.
        :param score: Score to limit the execution time.
        """
        super().__init__(limit=PyLimit.Score(score))


class GenerationsLimit(LimitBase):
    """
    Limit the number of generations of the engine.
    """

    def __init__(self, generations: int):
        """
        Initialize the generations limit.
        :param generations: Number of generations to limit the execution time.
        """
        super().__init__(limit=PyLimit.Generation(generations))

File: crates/radiate-python/src/old/registry/limit.rs
use crate::PyEngineParam;

const SECONDS_LIMIT: &str = "seconds";
const GENERATIONS_LIMIT: &str = "generations";
const SCORE_LIMIT: &str = "score";

pub enum Limit {
    Seconds(f64),
    Generations(usize),
    Score(f32),
}

impl Limit {
    pub fn seconds(seconds: f64) -> Self {
        Limit::Seconds(seconds)
    }

    pub fn generations(generations: usize) -> Self {
        Limit::Generations(generations)
    }

    pub fn score(score: f32) -> Self {
        Limit::Score(score)
    }
}

impl From<PyEngineParam> for Limit {
    fn from(param: PyEngineParam) -> Self {
        match param.name() {
            SECONDS_LIMIT => Limit::seconds(
                param
                    .get_args()
                    .get("seconds")
                    .map(|s| s.parse::<f64>().unwrap())
                    .unwrap_or(0.0),
            ),
            GENERATIONS_LIMIT => Limit::generations(
                param
                    .get_args()
                    .get("generations")
                    .map(|s| s.parse::<usize>().unwrap())
                    .unwrap_or(0),
            ),
            SCORE_LIMIT => Limit::score(
                param
                    .get_args()
                    .get("score")
                    .map(|s| s.parse::<f32>().unwrap())
                    .unwrap_or(0.0),
            ),
            _ => panic!("Invalid limit type"),
        }
    }
}
